File: lib/src/models/system_layout.dart
import 'package:meta/meta.dart';
// No collection needed for this specific file's direct classes, but good practice if they evolve.

@immutable
class SystemMargins {
  final double? leftMargin;
  final double? rightMargin;

  const SystemMargins({this.leftMargin, this.rightMargin});

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is SystemMargins &&
          runtimeType == other.runtimeType &&
          leftMargin == other.leftMargin &&
          rightMargin == other.rightMargin;

  @override
  int get hashCode => leftMargin.hashCode ^ rightMargin.hashCode;

  @override
  String toString() {
    return 'SystemMargins{leftMargin: $leftMargin, rightMargin: $rightMargin}';
  }
}

@immutable
class SystemDividers {
  final bool leftDivider;
  final bool rightDivider;

  const SystemDividers({this.leftDivider = false, this.rightDivider = false});

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is SystemDividers &&
          runtimeType == other.runtimeType &&
          leftDivider == other.leftDivider &&
          rightDivider == other.rightDivider;

  @override
  int get hashCode => leftDivider.hashCode ^ rightDivider.hashCode;

  @override
  String toString() {
    return 'SystemDividers{leftDivider: $leftDivider, rightDivider: $rightDivider}';
  }
}

@immutable
class SystemLayout {
  final SystemMargins? systemMargins;
  final double? systemDistance;
  final double? topSystemDistance;
  final SystemDividers? systemDividers;

  const SystemLayout({
    this.systemMargins,
    this.systemDistance,
    this.topSystemDistance,
    this.systemDividers,
  });

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is SystemLayout &&
          runtimeType == other.runtimeType &&
          systemMargins == other.systemMargins &&
          systemDistance == other.systemDistance &&
          topSystemDistance == other.topSystemDistance &&
          systemDividers == other.systemDividers;

  @override
  int get hashCode =>
      systemMargins.hashCode ^
      systemDistance.hashCode ^
      topSystemDistance.hashCode ^
      systemDividers.hashCode;

  @override
  String toString() {
    return 'SystemLayout{systemMargins: $systemMargins, systemDistance: $systemDistance, topSystemDistance: $topSystemDistance, systemDividers: $systemDividers}';
  }
}

File: MusicXMLParser/Models/SystemLayout.cs
using System;

namespace MusicXMLParser.Models
{
    public class SystemMargins
    {
        public double? LeftMargin { get; set; }
        public double? RightMargin { get; set; }

        public SystemMargins(double? leftMargin = null, double? rightMargin = null)
        {
            LeftMargin = leftMargin;
            RightMargin = rightMargin;
        }

        public override bool Equals(object obj)
        {
            if (obj is SystemMargins other)
            {
                return LeftMargin == other.LeftMargin &&
                       RightMargin == other.RightMargin;
            }
            return false;
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(LeftMargin, RightMargin);
        }

        public override string ToString()
        {
            return $"SystemMargins{{LeftMargin: {LeftMargin?.ToString() ?? "null"}, RightMargin: {RightMargin?.ToString() ?? "null"}}}";
        }
    }

    public class SystemDividers
    {
        public bool LeftDivider { get; set; }
        public bool RightDivider { get; set; }

        public SystemDividers(bool leftDivider = false, bool rightDivider = false)
        {
            LeftDivider = leftDivider;
            RightDivider = rightDivider;
        }

        public override bool Equals(object obj)
        {
            if (obj is SystemDividers other)
            {
                return LeftDivider == other.LeftDivider &&
                       RightDivider == other.RightDivider;
            }
            return false;
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(LeftDivider, RightDivider);
        }

        public override string ToString()
        {
            return $"SystemDividers{{LeftDivider: {LeftDivider}, RightDivider: {RightDivider}}}";
        }
    }

    public class SystemLayout
    {
        public SystemMargins SystemMargins { get; set; }
        public double? SystemDistance { get; set; }
        public double? TopSystemDistance { get; set; }
        public SystemDividers SystemDividers { get; set; }

        public SystemLayout(
            SystemMargins systemMargins = null,
            double? systemDistance = null,
            double? topSystemDistance = null,
            SystemDividers systemDividers = null)
        {
            SystemMargins = systemMargins;
            SystemDistance = systemDistance;
            TopSystemDistance = topSystemDistance;
            SystemDividers = systemDividers;
        }

        public override bool Equals(object obj)
        {
            if (obj is SystemLayout other)
            {
                return object.Equals(SystemMargins, other.SystemMargins) &&
                       SystemDistance == other.SystemDistance &&
                       TopSystemDistance == other.TopSystemDistance &&
                       object.Equals(SystemDividers, other.SystemDividers);
            }
            return false;
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(SystemMargins, SystemDistance, TopSystemDistance, SystemDividers);
        }

        public override string ToString()
        {
            return $"SystemLayout{{SystemMargins: {SystemMargins}, SystemDistance: {SystemDistance?.ToString() ?? "null"}, TopSystemDistance: {TopSystemDistance?.ToString() ?? "null"}, SystemDividers: {SystemDividers}}}";
        }
    }
}
